### PR TITLE
Reduce the number of applications API calls

### PR DESCRIPTION
**Pull request by agrare**
Query applications for supported application_types once per tenant
instead of once per source.

### CODE REVIEW

File: lib/topological_inventory/orchestrator/worker.rb [hunk 1]
```diff
@@ -165,6 +169,21 @@ def each_tenant
         each_resource(topology_internal_url_for("tenants")) { |tenant| yield tenant["external_tenant"] }
       end
 
+      # Set of ids for supported applications
+      def supported_application_type_ids
+        ids = []
+        each_resource(sources_api_url_for("application_types")) do |application_type|
```
Comment (bdunne) [inline lib/topological_inventory/orchestrator/worker.rb]: What do you think of letting the server handle the filtering with something like this?
```ruby
supported_applications_filter = SUPPORTED_APPLICATIONS.collect { |sa| "filter["name"][]=#{sa}" }
each_resource("#{sources_api_url_for("application_types")}?#{supported_applications_filter}").collect { |i| i["id"] }
```

File: spec/worker_spec.rb [hunk 2]
```diff
@@ -47,6 +47,31 @@
...
+      EOJ
+    end
+
+    let(:applications_response) do
+      <<~EOJ
+        {
+          "links": {},
+          "data": [
+            {"id":"1","application_type_id":"1","source_id":"1","tenant_id":"1"},
+            {"id":"1","application_type_id":"1","source_id":"2","tenant_id":"1"}
```
Comment (bdunne) [inline spec/worker_spec.rb]: "id" should be unique, right?

File: spec/worker_spec.rb [hunk 3]
```diff
@@ -120,56 +145,24 @@
...
-        }
-      EOJ
-    end
-
     it "generates the expected hash" do
       stub_rest_get("#{sources_api}/source_types", orchestrator_tenant_header, source_types_response)
+      stub_rest_get("#{sources_api}/application_types", orchestrator_tenant_header, application_types_response)
       stub_rest_get("http://topology.local:8080/internal/v1.0/tenants", orchestrator_tenant_header, tenants_response)
+
+      application_type_query = "filter%5Bapplication_type_id%5D%5Beq%5D%3D1%26filter%5Bapplication_type_id%5D%5Beq%5D%3D3"
```
Comment (bdunne) [inline spec/worker_spec.rb]: Should be `"filter%5Bapplication_type_id%5D%5Beq%5D%5B%5D%3D1%26filter%5Bapplication_type_id%5D%5Beq%5D%5B%5D%3D3"`

File: lib/topological_inventory/orchestrator/worker.rb (hunk 1 above)
Comment (bdunne) [reply lib/topological_inventory/orchestrator/worker.rb]: `supported_applications_filter` could even be a constant, right?

File: spec/worker_spec.rb (hunk 2 above)
Comment (agrare) [reply spec/worker_spec.rb]: D'oh yes it should

File: lib/topological_inventory/orchestrator/worker.rb [hunk 4]
```diff
@@ -165,6 +169,21 @@ def each_tenant
...
+          ids << application_type["id"] if SUPPORTED_APPLICATIONS.include?(application_type["name"])
+        end
+        ids
+      end
+
+      # URL to get a list of applications for supported application types
+      def supported_applications_url
+        query = CGI.escape(supported_application_type_ids.map { |id| "filter[application_type_id][eq]=#{id}" }.join("&"))
```
[the remark this answers is not in the document]
Comment (bdunne) [reply lib/topological_inventory/orchestrator/worker.rb]: [----] I, [2019-08-22T13:57:24.424518 #17778:1e3bd58]  INFO -- : Started GET "/api/v1.0/applications?filter[application_type_id][eq]=1&filter[application_type_id][eq]=3" for ::1 at 2019-08-22 13:57:24 -0400
[----] I, [2019-08-22T13:57:24.532207 #17778:1e3bd58]  INFO -- :   Parameters: {"filter"=>{"application_type_id"=>{"eq"=>"3"}}}

vs

[----] I, [2019-08-22T13:57:53.862452 #17778:1e3bc04]  INFO -- : Started GET "/api/v1.0/applications?filter[application_type_id][eq][]=1&filter[application_type_id][eq][]=3" for ::1 at 2019-08-22 13:57:53 -0400
[----] I, [2019-08-22T13:57:53.865636 #17778:1e3bc04]  INFO -- :   Parameters: {"filter"=>{"application_type_id"=>{"eq"=>["1", "3"]}}}

[the remark this answers is not in the document]
Comment (agrare) [reply lib/topological_inventory/orchestrator/worker.rb]: :+1:

File: lib/topological_inventory/orchestrator/worker.rb (hunk 1 above)
Comment (agrare) [reply lib/topological_inventory/orchestrator/worker.rb]: Yeah good idea, as for the `.collect` its going to have to wait for the enumerable PR :)

Comment (agrare) [reply lib/topological_inventory/orchestrator/worker.rb]: Hm this filter doesn't seem to work
`"filter[name][]=/insights/platform/catalog&filter[name][]=/insights/platform/topological-inventory"`

```
url = "http://localhost:3000/api/v1.0/application_types?filter%5Bname%5D%5B%5D%3D%2Finsights%2Fplatform%2Fcatalog%26filter%5Bname%5D%5B%5D%3D%2Finsights%2Fplatform%2Ftopological-inventory"
```

```
[----] I, [2019-08-22T15:03:03.100215 #20613:2b030fb17f00]  INFO -- : Started GET "/api/v1.0/application_types?filter%5Bname%5D%5B%5D%3D%2Finsights%2Fplatform%2Fcatalog%26filter%5Bname%5D%5B%5D%3D%2Finsights%2Fplatform%2Ftopological-inventory" for ::1 at 2019-08-22 15:03:03 -0400
[----] I, [2019-08-22T15:03:03.103222 #20613:2b030fb17f00]  INFO -- : Processing by Api::V1x0::ApplicationTypesController#index as */*
[----] I, [2019-08-22T15:03:03.103333 #20613:2b030fb17f00]  INFO -- :   Parameters: {"filter"=>{"name"=>[{"=/insights/platform/catalog&filter"=>{"name"=>[{"=/insights/platform/topological-inventory"=>nil}]}}]}}
[----] D, [2019-08-22T15:03:03.104479 #20613:2b030fb17f00] DEBUG -- :   Tenant Load (0.3ms)  SELECT  "tenants".* FROM "tenants" WHERE "tenants"."external_tenant" = $1 LIMIT $2  [["external_tenant", "system_orchestrator"], ["LIMIT", 1]]
[----] I, [2019-08-22T15:03:03.110783 #20613:2b030fb17f00]  INFO -- : Completed 500 Internal Server Error in 7ms (ActiveRecord: 1.8ms)


[----] F, [2019-08-22T15:03:03.111537 #20613:2b030fb17f00] FATAL -- :   
[----] F, [2019-08-22T15:03:03.111598 #20613:2b030fb17f00] FATAL -- : TypeError (can't cast ActionController::Parameters):
[----] F, [2019-08-22T15:03:03.111649 #20613:2b030fb17f00] FATAL -- :   
[----] F, [2019-08-22T15:03:03.111716 #20613:2b030fb17f00] FATAL -- : app/controllers/api/v1/mixins/index_mixin.rb:12:in `index'
app/controllers/application_controller.rb:52:in `block (2 levels) in with_current_request'
app/controllers/application_controller.rb:52:in `block in with_current_request'
app/controllers/application_controller.rb:46:in `with_current_request'
```

Comment (bdunne) [reply lib/topological_inventory/orchestrator/worker.rb]: Why `CGI.escape` over `URI.escape`?  The URI escaped string seems to work correctly.

test = "http://localhost:3000/stuff?filter[thing][]=a/b/c"

URI.escape(test)
=> "http://localhost:3000/stuff?filter[thing][]=a/b/c"

CGI.escape(test)
=> "http%3A%2F%2Flocalhost%3A3000%2Fstuff%3Ffilter%5Bthing%5D%5B%5D%3Da%2Fb%2Fc"

Comment (agrare) [reply lib/topological_inventory/orchestrator/worker.rb]: Rubocop would complain about using `URI.decode` and said to use `CGI.unescape` so I got in the habit of using CGI over URI.  Okay let me try that.

Comment (agrare) [reply lib/topological_inventory/orchestrator/worker.rb]: Okay that worked, interestingly the application filter worked but not the application_type filter.

**Pull request merged**